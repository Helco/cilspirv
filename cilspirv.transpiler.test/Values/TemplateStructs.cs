using System;
using NUnit.Framework;
using cilspirv.Library;
using cilspirv.Spirv;
using System.Linq;
using System.IO;

namespace cilspirv.transpiler.test.Values.Modules
{
    public class TemplateStructs
    {
        public struct Simple
        {
            [Location(0)] public float m;
        }

        [EntryPoint(ExecutionModel.Fragment)]
        [return: Output, Location(0)]
        public float RefFrag([Input] in Simple simple) => simple.m * 2;

        [EntryPoint(ExecutionModel.Vertex)]
        public void RefVert([Output] out Simple simple) => simple.m = 42;

        [EntryPoint(ExecutionModel.Fragment)]
        [return: Output, Location(0)]
        public float ValFrag([Input] Simple simple) => simple.m * 2;

        [EntryPoint(ExecutionModel.Vertex)]
        public void ValVert([Output] Simple simple) => simple.m = 42;

        [EntryPoint(ExecutionModel.Vertex)]
        [return: Output, Location(0)]
        public Simple ValReturnVert() => new() { m = 42 };
    }
}

namespace cilspirv.transpiler.test.Values
{
    public class TemplateStructs : ApprovalTranspileFixture<Modules.TemplateStructs>
    {
        [Test]
        public void RefShared() => VerifyModule(
            (nameof(Modules.TemplateStructs.RefFrag), isEntryPoint: true),
            (nameof(Modules.TemplateStructs.RefVert), isEntryPoint: true));

        [Test]
        public void ValShared() => VerifyModule(
            (nameof(Modules.TemplateStructs.ValFrag), isEntryPoint: true),
            (nameof(Modules.TemplateStructs.ValVert), isEntryPoint: true));

        /*[Test] initobj is not implemented, local vars can only be SPIRV
        public void ValReturnShared() => VerifyModule(
            (nameof(Modules.TemplateStructs.ValFrag), isEntryPoint: true),
            (nameof(Modules.TemplateStructs.ValReturnVert), isEntryPoint: true));*/
    }
}
