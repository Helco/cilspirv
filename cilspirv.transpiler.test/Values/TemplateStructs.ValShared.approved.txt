SPIRV 1.3
Generator 002A 1.0
OpCapability Shader
OpMemoryModel Logical Simple
OpEntryPoint Fragment #3 "ValFrag" #8 #7
OpEntryPoint Vertex #5 "ValVert" #9
OpExecutionMode #3 OriginUpperLeft
OpName #3 "ValFrag"
OpName #5 "ValVert"
OpName #7 "#return"
OpName #8 "System.Int32 cilspirv.transpiler.test.Values.Modules.TemplateStructs/Simple::m"
OpName #9 "System.Int32 cilspirv.transpiler.test.Values.Modules.TemplateStructs/Simple::m"
OpName #12 "System.Int32"
OpDecorate #7 Location
OpDecorate #8 Location
OpDecorate #9 Location
#12 = OpTypeInt 32 1
#13 = OpTypeVoid
#14 = OpTypeFunction #13
#15 = OpTypePointer Output #12
#16 = OpTypePointer Input #12
#10 = OpConstant #12 00000002
#11 = OpConstant #12 0000002A
#7 = OpVariable #15 Output 
#8 = OpVariable #16 Input 
#9 = OpVariable #15 Output 
#3 = OpFunction #13 None #14
#4 = OpLabel
#1 = OpLoad #12 #8 
#2 = OpIMul #12 #1 #10
OpStore #7 #2 
OpReturn
OpFunctionEnd
#5 = OpFunction #13 None #14
#6 = OpLabel
OpStore #9 #11 
OpReturn
OpFunctionEnd
